### PR TITLE
telemetry changes

### PR DESCRIPTION
**Pull request by andrewyuq**
1. UserTriggerDecision can now have Ignore and Unseen, (will later change to a new "MERGED" state indicating those other trigger results not accepted or rejected in the same display session)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt [hunk 1]
```diff
@@ -385,59 +390,60 @@ class CodeWhispererTelemetryService {
...
+
+                val previews = CodeWhispererService.getInstance().getAllSuggestionsPreviewInfo()
+                val recommendation =
+                    if (hasUserAccepted) {
+                        previews[sessionContext.selectedIndex].detail.recommendation
+                    } else {
+                        Completion.builder().content("").references(emptyList()).build()
+                    }
+                val referenceCount = if (hasUserAccepted && recommendation.hasReferences()) 1 else 0
+                val acceptedContent = if (hasUserAccepted) recommendation.content() else ""
```
Comment (Will-ShaoHua) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:424]: think the if else here is extra, could be `val acceptedContent = recommendation.content`

Comment (andrewyuq) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:424]: in rejected cases we want the acceptedContent to be "", or at least this is the existing behavior.

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt [hunk 2]
```diff
@@ -385,59 +390,60 @@ class CodeWhispererTelemetryService {
...
+                val generatedLineCount = if (acceptedContent.isEmpty()) 0 else acceptedContent.split("\n").size
```
[the remark this answers is not in the document]
Comment (andrewyuq) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:425]: `"".split("\n").size` is 1

[the remark this answers is not in the document]
Comment (Will-ShaoHua) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:425]: isee

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt (hunk 1 above)
Comment (Will-ShaoHua) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:424]: i mean you already had this logic here, isn't it
```
val recommendation =
                    if (hasUserAccepted) {
                        previews[sessionContext.selectedIndex].detail.recommendation
                    } else {
                        Completion.builder().content("").references(emptyList()).build()
                    }
```

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt [hunk 3]
```diff
@@ -385,59 +390,60 @@ class CodeWhispererTelemetryService {
...
-                sessionContext.seen.contains(index),
-                hasUserAccepted,
-                detailContext.isDiscarded,
-                detailContext.recommendation.content().isEmpty()
-            )
-            sendUserDecisionEvent(requestContext, responseContext, detailContext, index, suggestionState, detailContexts.size)
+        CodeWhispererService.getInstance().getAllPaginationSessions().forEach { (jobId, state) ->
+            if (state == null) return@forEach
+            val decisions = mutableListOf<CodewhispererSuggestionState>()
+            val details = state.recommendationContext.details
```
[the remark this answers is not in the document]
Comment (andrewyuq) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:400]: technically we can but we are still going to call `CodeWhispererService.getInstance().getAllPaginationSessions()` somewhere in the caller.

[the remark this answers is not in the document]
Comment (Will-ShaoHua) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:400]: okk, but we need some mechanism to ensure the states here is correct and won't have dirty data

[the remark this answers is not in the document]
Comment (Will-ShaoHua) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:400]: when will the states in codehispererService get reset?

[the remark this answers is not in the document]
Comment (andrewyuq) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:400]: in disposeDisplaySession()

[the remark this answers is not in the document]
Comment (andrewyuq) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:400]: everytime when a display session ends

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt (hunk 1 above)
Comment (andrewyuq) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:424]: yes, this is correct, I forgot I put a recommendation variable there, let me change this

Comment (andrewyuq) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt]: fixed

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (Will-ShaoHua) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:400]: gotcha, thanks for the explanation

**Review by Will-ShaoHua — APPROVED** [review]

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrewyuq) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:400]: so here's the current flow:
1. `CodeWhispererTelemetryService.getInstance().sendUserDecisionEventForAll()` is only called when we dispose the sessionContext.
2. We dispose the sessionContext only when we call `CodeWhispererService.getInstance().disposeDisplaySession()`
3. So before disposing the session, `CodeWhispererService.getInstance().getAllPaginationSessions()` holds the trigger/suggestions info which is up-to-date. And we immediately clear the `ongoingRequests` map after the telemetry has been sent to make sure a fresh new session context on a new display session.

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt [hunk 4]
```diff
@@ -385,59 +390,60 @@ class CodeWhispererTelemetryService {
...
 
-            decisions.add(suggestionState)
-        }
+            details.forEachIndexed { index, detail ->
```
Comment (rli) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt:402]:
```kt
val decisions = details.mapIndexed {
                val suggestionState = recordSuggestionState(detail, hasUserAccepted)
                sendUserDecisionEvent(state.requestContext, state.responseContext, detail, index, suggestionState, details.size)

                suggestionState
}
```
?

**Review by rli — APPROVED** [review]

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt (hunk 4 above)
Comment (andrewyuq) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryService.kt]: addressed

**Pull request merged**